src/material/tree/testing/tree-harness.ts

```diff
@@ -28,4 +33,103 @@ export class MatTreeHarness extends ComponentHarness {
   async getNodes(filter: TreeNodeHarnessFilters = {}): Promise<MatTreeNodeHarness[]> {
     return this.locatorForAll(MatTreeNodeHarness.with(filter))();
   }
+
+  /**
+   * Object representation of the visible tree structure.
+   * If a node is under an unexpanded node it will not be shown here.
+   * Eg.
+   * Tree (all nodes expanded):
+   * `
+   * <mat-tree>
+   *   <mat-tree-node>Node 1<mat-tree-node>
+   *   <mat-nested-tree-node>
+   *     Node 2
+   *     <mat-nested-tree-node>
+   *       Node 2.1
+   *       <mat-tree-node>
+   *         Node 2.1.1
+   *       <mat-tree-node>
+   *     <mat-nested-tree-node>
+   *     <mat-tree-node>
+   *       Node 2.2
+   *     <mat-tree-node>
+   *   <mat-nested-tree-node>
+   * </mat-tree>`
+   *
+   * Tree structure:
+   * {
+   *  children: [
+   *    {
+   *      text: 'Node 1',
+   *      children: [
+   *        {
+   *          text: 'Node 2',
+   *          children: [
+   *            {
+   *              text: 'Node 2.1',
+   *              children: [{text: 'Node 2.1.1'}]
+   *            },
+   *            {text: 'Node 2.2'}
+   *          ]
+   *        }
+   *      ]
+   *    }
+   *  ]
+   * };
+   */
+  async getTreeStructure(): Promise<TreeNode> {
+    const nodes = await this.getNodes();
+    const nodeInformation = await parallel(() => nodes.map(node => {
+        return Promise.all([node.getLevel(), node.getText(), node.isExpanded()]);
+      }));
+    return this._getTreeStructureRecursion(nodeInformation, 1, true);
+  }
+
+  private _getTreeStructureRecursion(tree: [number, string, boolean][],
+                                     level: number,
+                                     parentExpanded: boolean): TreeNode {
+    // param level is the level of nodes that are being accounted for during this iteration
```

Comment:
This might be better expressed as a doc string on the parameter:

```ts
private _getTreeStructureRecursion(
    /** ... */
    tree: [number, string, boolean][],
    /** The level of nodes that are being accounted for during this iteration */
    level: number,
```